packages/ra-core/src/routing/useGetPathForRecordCallback.ts

```diff
@@ -41,19 +44,41 @@ export const useGetPathForRecordCallback = <
             const linkResultIsPromise = isPromise(linkResult);
 
             if (linkResultIsPromise) {
-                return linkResult.then(resolvedLink => {
-                    if (resolvedLink === false) {
-                        // already set to false by default
-                        return;
+                const resolvedLink = await linkResult;
+                if (resolvedLink === false) {
+                    // already set to false by default
+                    return;
+                }
+                if (['edit', 'show'].includes(resolvedLink)) {
+                    if (
+                        !(await canAccess({
+                            action: resolvedLink,
```

Comment:
`resolvedLink` is like `/posts/123`, it's not `edit` or `show`, am I right? If so, it's not what you want to use for the `canAccess` action.

Reply:
Actually no, not here. We use `resolvedLink` as the `type` for `createPath` below

Reply:
Besides, I think we should actually check for both `edit` and `show` when both are available as you may have access to one but not the other and that may change the `resolvedLink`

Reply:
As discussed, this is very likely not a case

Reply:
I don't think we should do the access control check here, as we don't know if the link is a default one or one specified by the developer. If it's the default, it's the responsibility of this hook to determine a default compatible with access control. If it's one specified by the developer, it's their responsibility and we should not do any access control. 

Instead, you should check whether the user has access to show and edit views preemptively, and use the result when computing defaultLink above.